notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "isolated-drove",
+   "metadata": {},
+   "source": [
+    "$$\n",
+    "\\partial z^*(a) = -[\\partial_1 f(a_0,z_0))]^{-1} \\partial_0 f(a_0,z_0))\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "agricultural-rebel",
+   "metadata": {},
+   "source": [
+    "This expression can be efficiently solved. If you are interested in a derivation and how one could implement this in `jax` then I can highly recommend chapter $2$ of the NeurIPS 2020 [Deep Implicit Layers](https://implicit-layers-tutorial.org/) tutorial.\n",
+    "\n",
+    "In order to use implicit differentiation we first need to rewrite our optimization problem as a root finding procedure. This is easily done since the force is $0$ at the energy minimum.\n",
+    "\n",
+    "I'll first define a new function that uses implicit differentiation and then I'll highlight the differences."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "consistent-singing",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Let's go back to our small system in order to compare explicit and implicit differentiation.\n",
+    "N = 7\n",
+    "density = 0.3\n",
+    "\n",
+    "dimension = 2\n",
+    "box_size = quantity.box_size_at_number_density(N,density,dimension)\n",
+    "displacement, shift = space.periodic(box_size) \n",
+    "\n",
+    "key = random.PRNGKey(5)\n",
+    "key, split = random.split(key)\n",
+    "R_init = random.uniform(key, (N,dimension), minval=0.0, maxval=box_size, dtype=f64) \n",
+    "\n",
+    "sigma = jnp.full((N,N), 2.0)\n",
+    "alpha = jnp.full((N,N),2.)\n",
+    "param_dict = {'sigma':sigma,'alpha':alpha}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "mature-discharge",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def implicit_diff(params,R_init,displacement):\n",
+    "    energy_fn = energy.soft_sphere_pair(displacement, **params)\n",
+    "    force_fn = jit(quantity.force(energy_fn))\n",
+    "    \n",
+    "    # wrap force_fn with a lax.stop_gradient to prevent a CustomVJPException\n",
+    "    no_grad_force_fn = jit(lambda x: lax.stop_gradient(force_fn(x)))\n",
+    "    \n",
+    "    # make the dependence on the variables we want to differentiate explicit\n",
+    "    explicit_force_fn = jit(lambda R, p: force_fn(R, **p))\n",
+    "    \n",
+    "    def solver(params, x):\n",
+    "        # params are unused\n",
+    "        del params\n",
+    "        # need to use no_grad_force_fn!\n",
```

Comment:
Here and elsewhere can you use proper grammar for comments (capitalization and punctuation).